presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java

```diff
@@ -17,24 +17,28 @@
 
 import javax.annotation.concurrent.ThreadSafe;
 
-import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.LongConsumer;
 
+import static com.google.common.base.MoreObjects.toStringHelper;
 import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
 @ThreadSafe
-class PageReference
+final class PageReference
 {
+    private static final AtomicIntegerFieldUpdater<PageReference> REFERENCE_COUNT_UPDATER = AtomicIntegerFieldUpdater.newUpdater(PageReference.class, "referenceCount");
```

Comment:
I don't think `AtomicInteger` is a problem. Page granularity is not such a big deal. If there was `AtomicInteger` per row that would be a problem

Reply:
It's still an extra allocation (with object headers and alignment overhead) and extra pointer indirection at usage time. The improvement is minor but enough of these are created through local exchanges that embedding the counter is worthwhile.